### ISSUE TITLE
"No locks available." error while boostrapping a YCM superbuild

### ISSUE THREAD

**Issue by traversaro** [robotology/ycm#93]
While configuring the `codyco-superbuild` on the shared disk (NFS) of the new iCubGenova05 setup running Ubuntu 16.04 , I get this error: 

```
 CMake Error at cmake/IncludeUrl.cmake:241 (file):
   error locking file


 "/usr/local/src/robot/codyco-superbuild/build/CMakeFiles/YCMEPHelper.cmake.cma
 ke"

   No locks available.
 Call Stack (most recent call first):
   cmake/YCMBootstrap.cmake:105 (include_url)
   CMakeLists.txt:84 (include)
```

Placing the build in the "normal" filesystem running ext4 solves the problem. 
I guess the error is not directly related to ycm (probably the default configuration for the NFS server in Ubuntu 16.04 are different the lock server is not running by  default) but it is worth opening an issue so future users with the same problem will now how to solve the problem.

**Comment by mbrunettini** [comment]
@traversaro : can you explain hot to reproduce thi issue?
So I can make some tests on my own.
Thanks

**Comment by traversaro** [comment]
On the `iCubGenova05` setup: 
~~~
cd /usr/local/robot/src/codyco-superbuild/build
cmake . 
~~~

**Comment by drdanz** [comment]
From `IncludeUrl.cmake` line 240

```cmake
  if(NOT CMAKE_VERSION VERSION_LESS 3.2)
    file(LOCK "${_lockFile}")
  endif()
```

I'd say it that `file(LOCK)` was never tested on NFS.

The check on the CMake version explains why it was working before, but it fails on 16.04 that has CMake 3.5.1

**Comment by traversaro** [comment]
For the time being I commented out all the `file(LOCK ...)` in the codyco-superbuild `IncludeUrl.cmake`, as I don't think that conflicts in downloading superbuild files are a real problem, while NFS compatibility is definitely a big requirement.

**Comment by mbrunettini** [comment]
There are mainly two methods to lock a file:
- [flock](http://man7.org/linux/man-pages/man2/flock.2.html)
- [lockf](http://man7.org/linux/man-pages/man3/lockf.3.html)
While since Linux 2.6.12, NFS clients support flock(), lockf does not work over NFS.
See this [page](http://unix.stackexchange.com/questions/229675/nfs-file-locking-not-working-am-i-misunderstanding)

I don't know, which one cmake uses, but i suspect flock()

**Comment by traversaro** [comment]
Reopened because this bug in YCM is still not solved: the bootstrap fails on disk mounted using the NFS.

**Comment by traversaro** [comment]
Just a remainder about the proper solution: the `file(LOCK "${_lockFile}")` return an error code: if the error is because the FS does not support locking, it should be ignored, while the system should fail only if the FS supports the locking and the file is actually locked.

**Comment by drdanz** [comment]
I'm not sure how complicate is to parse the output of `file(LOCK)`, and if this output is likely to change.

As an alternative, we could just have an option to disable locks...

Or we could run something like this

```cmake
  unset(_lock_fs)
  if(UNIX)
    get_filename_component(_localDir ${_localFile} DIRECTORY)
    if(EXISTS "${_localDir}")
      execute_process(COMMAND df -T ${_localDir}
                      OUTPUT_VARIABLE _df_out
                      OUTPUT_STRIP_TRAILING_WHITESPACE)
      string(REGEX REPLACE ".+\n" "" _df_out ${_df_out})
      string(REGEX REPLACE " +" ";" _df_out ${_df_out})
      list(GET _df_out 1 _lock_fs)
    endif()
  endif()

#  [...]

  if(NOT CMAKE_VERSION VERSION_LESS 3.2 AND NOT _lock_fs MATCHES "nfs")

# [...]
```

**Comment by traversaro** [comment]
IMHO parsing the output of `RESULT_VARIABLE` seems to be more robust then launching an external command, because we just need to track the CMake implementation of `file(LOCK ..)`.

**Comment by drdanz** [comment]
@traversaro I agree, but maybe I didn't clarify my thoughts... I'm worried that the output could be the output received from the internal command used, and that this output could be dependent on the system on which cmake is running and not on the `file(LOCK)` implementation.

**Comment by drdanz** [comment]
@traversaro Maybe another option could also be to move the lock file to /run/lock instead of keeping it in the build directory? I'm not sure if keeping it outside of the build directory is a good idea though...

**Comment by traversaro** [comment]
> @traversaro Maybe another option could also be to move the lock file to /run/lock instead of keeping it in the build directory? I'm not sure if keeping it outside of the build directory is a good idea though...

I do not think it make sense, the lock is used to make sure that two instances of CMake do not call IncludeUrl on the same file at the same time, while it is totally ok for two instance of CMake to download two different file using IncludeUrl at the same time. 

To be honest, I do not think thread safety when running two instance of cmake is a bit problem, and probably running two instance of cmake in the same superbuild build fails for other reasons, so I would simply drop the `file(LOCK)`.

**Comment by drdanz** [comment]
> I would simply drop the `file(LOCK)`.

That probably makes sense... I don't remember why I used to like it at the time, but it does not make sense anymore to me.

**Comment by drdanz** [comment]
Fixed in #228
